packages/amplify-graphql-generation-transformer/src/grapqhl-generation-transformer.ts

```diff
@@ -0,0 +1,219 @@
+import { GenerationDirective } from '@aws-amplify/graphql-directives';
+import {
+  DirectiveWrapper,
+  InvalidDirectiveError,
+  TransformerPluginBase,
+  generateGetArgumentsInput,
+  TransformerResolver,
+} from '@aws-amplify/graphql-transformer-core';
+import {
+  MappingTemplateProvider,
+  TransformerContextProvider,
+  TransformerSchemaVisitStepContextProvider,
+} from '@aws-amplify/graphql-transformer-interfaces';
+import { DirectiveNode, FieldDefinitionNode, InterfaceTypeDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
+import { HttpResourceIDs, ResolverResourceIDs } from 'graphql-transformer-common';
+import { ToolConfig, createResponseTypeTool } from './utils/tools';
+import * as cdk from 'aws-cdk-lib';
+import { createInvokeBedrockResolverFunction } from './resolvers/invoke-bedrock';
+import * as iam from 'aws-cdk-lib/aws-iam';
+import { Construct } from 'constructs';
+import { validate } from './validation';
+
+export type GenerationDirectiveConfiguration = {
+  parent: ObjectTypeDefinitionNode;
+  directive: DirectiveNode;
+  aiModel: string;
+  field: FieldDefinitionNode;
+  systemPrompt: string;
+  inferenceConfiguration: InferenceConfiguration | undefined;
+};
+
+export type InferenceConfiguration = {
+  maxTokens?: number;
+  temperature?: number;
+  topP?: number;
+};
+
+export type GenerationConfigurationWithToolConfig = GenerationDirectiveConfiguration & {
+  toolConfig: ToolConfig;
+};
+
+export class GenerationTransformer extends TransformerPluginBase {
+  private directives: GenerationDirectiveConfiguration[] = [];
+
+  constructor() {
+    super('amplify-generation-transformer', GenerationDirective.definition);
+  }
+
+  field = (
+    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
+    definition: FieldDefinitionNode,
+    directive: DirectiveNode,
+    context: TransformerSchemaVisitStepContextProvider,
+  ): void => {
+    const directiveWrapped = new DirectiveWrapper(directive);
+    const config = directiveWrapped.getArguments(
+      {
+        parent,
+        directive,
+        field: definition,
+        inferenceConfiguration: {},
+      } as GenerationDirectiveConfiguration,
+      generateGetArgumentsInput(context.transformParameters),
+    );
+
+    validate(config, context as TransformerContextProvider);
+    this.directives.push(config);
+  };
+
+  generateResolvers = (ctx: TransformerContextProvider): void => {
+    if (this.directives.length === 0) return;
+
+    this.directives.forEach((directive) => {
+      const { parent, field } = directive;
+      const fieldName = field.name.value;
+      const parentName = parent.name.value;
+
+      const directiveWithToolConfig: GenerationConfigurationWithToolConfig = {
+        ...directive,
+        toolConfig: createResponseTypeTool(field, ctx),
+      };
+
+      const stackName = `Generation${this.capitalizeFirstLetter(fieldName)}BedrockDataSourceStack`;
+      const stack = this.createStack(ctx, stackName);
```

Comment:
> E.g., should we create a stack to house a data source & invocation function for each model type, which is then referenced by other model fields? which is then referenced by other model fields?

I'm not sure what you mean with "referenced by other model fields" here. Can you clarify? 

The nested stack is created for each generation route because each has it's own data source and resolver. If "models" refers to `@model` types, those aren't related to anything happening in deployment of the generation route resources.

Reply:
Yeah, I was forgetting that models aren't strongly associated with these operations, so let's forget I mentioned that. :)

I'm assuming that _**each**_ generation route will create the following stack resources:
- An AppSync pipeline resolver, containing the following AppSync function resources:
  - begin
  - preauth
  - postauth
  - data
- An AppSync HTTP datasource for making the underlying bedrock API call
- An IAM policy scoped to allow AppSync to access Bedrock
- ...what else?

> The nested stack is created for each generation route because each has it's own data source and resolver

This is where I'm focusing. Why does each route need its own datasource and IAM policy? Looking at this, we could optimize resource usage by abstracting the HTTP datasource and IAM policies into a common resource, that would be reuseable across generation routes. As an HTTP data source, there really isn't anything to configure beyond the endpoint, region, signing service name. The request generated by the resolver would contain details like model name, system prompt, schema, etc.

Basically, I'm looking for the impact this will have on a customer's stack resource count, and opportunities to optimize, before we get into a situation where customers have deployed dozens of routes, each with at least 2 stack resources that aren't actually specific to that route.

Also, tools might add another handful of resources to the stack -- would they be reusable across queries?

Reply:
Ah ok, I see. Thanks for that context.

I actually considered doing what you described, but it _felt_ wrong because there wasn't a clean mapping of route to datasource, which (again) _felt_ like it could get messy in this valid and expected scenario:
### Initial deployment
```ts
generateFoo: a.generation({ 
  aiModel: a.ai.model('Model A'), // example model, doesn't exist
  systemPrompt: "...",
})
// other stuff ...

generateBar: a.generation({ 
  aiModel: a.ai.model('Model A'),
  systemPrompt: "...",
})
// other stuff ...
```
#### Resources
- BedrockModelADataSource
- Resolver x 2 (one for each route)
- IAM Role with inline policy

### Subsequent Deployment
```ts
generateFoo: a.generation({ 
  aiModel: a.ai.model('Model A'),
  systemPrompt: "...",
})
// other stuff ...

generateBar: a.generation({ 
  aiModel: a.ai.model('Model B'),
  systemPrompt: "...",
})
// other stuff ...
```

#### Resources
- BedrockModelADataSource
- BedrockModelBDataSource
- Resolver x 2 (one for each route)
- IAM Role with inline policy x 2 (one for each data source)


The part that _feels_ off to me is that a change in model can result in a new data source deployment. I'd be very happy to be told that my fear is just ignorance, and that this approach actually makes sense.

> Also, tools might add another handful of resources to the stack -- would they be reusable across queries?
Nope. Tools are just an object defined within the resolver code, no additional resources. 

Here's an example stack deployed:
<img width="2017" alt="Screenshot 2024-08-30 at 5 04 30 PM" src="https://github.com/user-attachments/assets/fc7b6972-67a3-4b35-aae9-5b314da13e59">

Reply:
> I'm assuming that each generation route will create the following stack resources:
> 
>     An AppSync pipeline resolver, containing the following AppSync function resources:
>         begin
>         preauth
>         postauth
>         data

Just `auth` and `data`.

Reply:
I think Tim has a good point here and might be worth discussing the trade offs with team before we release. I couldn't find any AppSync imposed limit on number of datasources, but feel like re-use will save us from trouble in the future.